cmd/auth-provider-gcp/gcpcredential/gcpcredential.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// Enabled implements DockerConfigProvider for all of the Google implementations.
+func (g *MetadataProvider) Enabled() bool {
+	return onGCEVM()
+}
+
+// Provide implements DockerConfigProvider
+func (g *DockerConfigKeyProvider) Provide(image string) credentialconfig.DockerConfig {
+	// Read the contents of the google-dockercfg metadata key and
+	// parse them as an alternate .dockercfg
+	if cfg, err := credentialconfig.ReadDockerConfigFileFromURL(DockerConfigKey, g.Client, metadataHeader); err != nil {
+		klog.Errorf("while reading 'google-dockercfg' metadata: %v", err)
+	} else {
+		return cfg
+	}
+
+	return credentialconfig.DockerConfig{}
+}
+
+// Provide implements DockerConfigProvider
+func (g *DockerConfigURLKeyProvider) Provide(image string) credentialconfig.DockerConfig {
+	// Read the contents of the google-dockercfg-url key and load a .dockercfg from there
+	if url, err := credentialconfig.ReadURL(DockerConfigURLKey, g.Client, metadataHeader); err != nil {
+		klog.Errorf("while reading 'google-dockercfg-url' metadata: %v", err)
+	} else {
+		if strings.HasPrefix(string(url), "http") {
+			if cfg, err := credentialconfig.ReadDockerConfigFileFromURL(string(url), g.Client, nil); err != nil {
+				klog.Errorf("while reading 'google-dockercfg-url'-specified url: %s, %v", string(url), err)
+			} else {
+				return cfg
+			}
+		} else {
+			// TODO(mattmoor): support reading alternate scheme URLs (e.g. gs:// or s3://)
+			klog.Errorf("Unsupported URL scheme: %s", string(url))
+		}
+	}
+
+	return credentialconfig.DockerConfig{}
+}
+
+// runWithBackoff runs input function `f` with an exponential backoff.
+// Note that this method can block indefinitely.
+func runWithBackoff(f func() ([]byte, error)) []byte {
+	var backoff = 100 * time.Millisecond
+	const maxBackoff = time.Minute
+	for {
+		value, err := f()
+		if err == nil {
+			return value
+		}
+		time.Sleep(backoff)
+		backoff = backoff * 2
+		if backoff > maxBackoff {
+			backoff = maxBackoff
+		}
+	}
+}
+
+// Enabled implements a special metadata-based check, which verifies the
+// storage scope is available on the GCE VM.
+// If running on a GCE VM, check if 'default' service account exists.
+// If it does not exist, assume that registry is not enabled.
+// If default service account exists, check if relevant scopes exist in the default service account.
+// The metadata service can become temporarily inaccesible. Hence all requests to the metadata
+// service will be retried until the metadata server returns a `200`.
+// It is expected that "http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/" will return a `200`
+// and "http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/default/scopes" will also return `200`.
+// More information on metadata service can be found here - https://cloud.google.com/compute/docs/storing-retrieving-metadata
+func (g *ContainerRegistryProvider) Enabled() bool {
+	if !onGCEVM() {
```

Comment:
Not sure we need this check anymore.

Reply:
When this gets copied/merged back into the main repository, the legacy credential provider interface will still use this code, I think (since this comes from https://github.com/kubernetes/kubernetes/pull/95775 as well). I could definitely remove it (we won't ever need it in cloud-provider-gcp), but I thought it might be good to point that out.

Reply:
Fixed